### PR TITLE
NFC: Upgrade the Flow dialect to use tablegen-based pass declarations.

### PR DESCRIPTION
**Pull request by stellaraccident**
* Eliminates static registration from the flow dialect.
* While a good idea to do everywhere, this was done while triaging a startup crash in release mode with assertions enabled that appeared to be static initializer based (it was not - and now still happens at registration time).
* We should do this to all passes in the system eventually (but it would be nice to delete old cruft first).

### CODE REVIEW

**Comment by stellaraccident** [comment]
River mentions that we can reduce some bits further by defining statistics in tablegen: https://github.com/llvm/llvm-project/blob/cab3c6c6c48ed0da0f047bdc1a337d991c78ad8c/mlir/include/mlir/Transforms/Passes.td#L378

File: iree/compiler/Dialect/Flow/Transforms/PassDetail.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+#define GEN_PASS_CLASSES
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h.inc"
```
Comment (GMNGeoffrey) [inline iree/compiler/Dialect/Flow/Transforms/PassDetail.h:26]: So I know that MLIR never does this, but shouldn't we be undefing this macro after we're done with it?

```suggestion
#define GEN_PASS_CLASSES
#include "iree/compiler/Dialect/Flow/Transforms/Passes.h.inc"
#undef GEN_PASS_CLASSES
```

(same throughout)

**Review by GMNGeoffrey — APPROVED** [review]
Overall LGTM with some nits

File: iree/compiler/Dialect/Flow/Transforms/PassDetail.h (hunk 1 above)
Comment (stellaraccident) [reply iree/compiler/Dialect/Flow/Transforms/PassDetail.h:26]: Actually, the undef happens consistently in the generated code, so not going to commit this one.

Comment (GMNGeoffrey) [reply iree/compiler/Dialect/Flow/Transforms/PassDetail.h:26]: Ah! That makes sense

**Pull request merged**